File: backend/src/xiangqi/move.ts
import { BoardPoint } from './boardPoint';
import { Piece } from './piece';

export interface Move {
  readonly from: BoardPoint;
  readonly to: BoardPoint;
  readonly movedPiece: Piece;
  readonly capturedPiece: Piece | null;
}

export function isCapture(move: Move): boolean {
  return move.capturedPiece !== null;
}

File: lib/domain/models/move.dart
import 'board_point.dart';
import 'piece.dart';

/// A single, already-executed-or-proposed piece movement.
class Move {
  final BoardPoint from;
  final BoardPoint to;
  final Piece movedPiece;
  final Piece? capturedPiece;

  const Move({
    required this.from,
    required this.to,
    required this.movedPiece,
    this.capturedPiece,
  });

  bool get isCapture => capturedPiece != null;

  @override
  String toString() => '$movedPiece $from->$to${isCapture ? ' x$capturedPiece' : ''}';
}
